docs/design/2020-05-08-standardize-error-codes-and-messages.md

```diff
@@ -0,0 +1,166 @@
+    },
+    {
+        "code": 8110,
+        "error": "The Cartesian product operation cannot be executed.",
+        "message": "Set cross-join in the configuration to true.\nThe second line."
+    }
+]
+```
+
+Benefit:
+- It's easy to write (I think every developer knows json).
+- It's easy to parse (I think every programming language can handle json).
+- It's easy to implement (parse json and render markdown content)
+
+Tradeoff:
+- Poor readability when multiple lines are involved.
+- The markdown content can't be previewed on writing.
+
+#### Meta file in toml
+
+```toml
+[[errors]]
+code = 8005
+error = '''Transactions in TiDB encounter write conflicts.'''
+message = '''See [the Troubleshoot section](https://pingcap.com/docs/stable/faq/tidb#troubleshoot) for the cause and solution.'''
+
+[[errors]]
+code = 8110
+error = '''The Cartesian product operation cannot be executed.'''
+message = '''
+Set cross-join in the configuration to true.
+The second line.
+'''
+```
+
+Benefit:
+- It's easy to write.
+- It handles multiple lines well.
+- It's easy to implement (parse toml and render markdown content)
+
+Tradeoff:
+- The markdown content can't be previewed on writing.
+
+#### Meta file in markdown
+
+```md
+## Code: 8005
+### Error
+Transactions in TiDB encounter write conflicts.
+### Message
+See [the Troubleshoot section](https://pingcap.com/docs/stable/faq/tidb#troubleshoot) for the cause and solution.
+
+## Code: 8110
+### Error
+The Cartesian product operation cannot be executed.
+### Message
+Set cross-join in the configuration to true.
+The second line.
+```
+
+Benefit:
+- It's easy to write.
+- It handles multiple lines well.
+- The markdown content can be previewed on writing.
+
+Tradeoff:
+- It's hard to implement (see example below)
+- It's not self-consistent (see example below)
+
+Tradeoff Example:
+
+```md
+## Code: 8110
+### Error
+The Cartesian product operation cannot be executed.
+### Message
+## Code: 8005
+### Error
+Transactions in TiDB encounter write conflicts.
+### Message
+See [the Troubleshoot section](https://pingcap.com/docs/stable/faq/tidb#troubleshoot) for the cause and solution.
+```
+
+As the syntax above, the 8005 block is the message part of 8110 block, so we expect it's result is the same as this toml:
+
+```toml
+[[errors]]
+code = 8110
+error = '''The Cartesian product operation cannot be executed.'''
+message = '''
+## Code: 8005
+### Error
+Transactions in TiDB encounter write conflicts.
+### Message
+See [the Troubleshoot section](https://pingcap.com/docs/stable/faq/tidb#troubleshoot) for the cause and solution.
+'''
+```
+
+In this case, we must define the grammar and write a parser. Writing parser increase much complexity. What's worse
+is that I found no grammar that fits it.
+
+### Principle
+
+In every build, the pipeline should fetch all these meta files from all repositories:
+
+```bash
+mkdir -p errors
+curl https://raw.githubusercontent.com/pingcap/tidb/master/errors.json -o errors/tidb.json
```

Comment:
How about a direct parsing of https://github.com/pingcap/docs/blob/master/reference/error-codes.md? This only requires parsing one file.

Reply:
This file is the target file we will generate from metafiles, not the source file.

Because if we use a separated repo to record error code, every time new error code is added, the developer who introduced the error must give two PRs, one modify code, one modify doc.